### PR TITLE
Adds data models / new API Client

### PR DESCRIPTION
**Pull request by AngellusMortis**
After doing a lot of thinking of implementation, I decided to go a bit of a different path for adding data models. Trying to retrofit the new data models into the existing API class feels like it would add a signfanctly amount of overhead and potentially cause a slowdown on performance. So I have decided to make a 100% seperate API class (`ProtectAPIClient`) that we can migrate too when it has feature parity with the existing one. This ensures backwards compatibility without any of the overhead.

## Overview of changes
* Adds models for all of the data types from Unifi Protect under `pyunifiprotect.data`
* Adds a separate `ProtectApiClient` which will be a separate, new and breaking change to the existing `UpvServer` API class.
    * The new API client only really does updates as of now. The `update` method is on par with the existing one and it will update from the events stream or the Websocket accordingly.
    * The Websocket / event code works signfanctly differently now. The `.bootstrap` is essentially the replacement for `.devices` and now has real objects representing the devices. This is updated in real-time as new events / ws messages come. This better mirrors how the Unifi Protect works in the browser.
* Adds CLI option to the `shell` command to use the new API Client class.
* Removes tests that were design for testing backwards compat breaking changes for `UpvServer` since I am building a separate API class.

## Next Steps (after this PR)

* Following the discussion we had before, the next PR would be to make it so we can use `pydantic` without enforcing runtime validation in case we need to turn it off if it is causing issues.
* Add Websocket subscriptions
* Adds the various missing `get_` methods from the existing  `UpvServer` API class
* Add saving/updating devices to the data models / API class

## HA Install Link:

```
git+https://github.com/AngellusMortis/pyunifiprotect.git@feature/data-models#pyunifiprotect==0.35.0
```

### CODE REVIEW

**Comment by bdraco** [comment]
I'm a big fan of enforcing type hints in the CI.

Doing it at run time seems a bit worrisome given that the unifi api isn't published and ubiquity is likely to make future changes between versions. I think the risk of breakage is higher than other projects that use `pydantic`. I'd feel quite differently if ubiquity wants to provide a proper published and versioned api.

It is @briis 's call to make though.

**Comment by AngellusMortis** [comment]
I do agree it is very dependent on how "stable" the API. @briis would have the most insight into that since he has likely supported this the longest.

Generally though, I think any breaking change would likely still cause major issues. Enforcing the runtime checking only makes of aware of those issues ASAP, though it does eliminate any chance of the application to continue to work with a degraded state.

**Comment by AngellusMortis** [comment]
It is worth mentioning that there _is_ a way to downgrade Unifi Protect. Though it is not officially supported. I just used it to downgrade to 1.19.0 without any noticeable problems.

(not sure of the process if you are not using unifi-os):

```
ssh root@udm-pro-ip
unifi-os shell
apt-get update
apt-get install --reinstall --allow-downgrades unifi-protect=1.19.0 -y
```

Gotten from: https://community.ui.com/releases/UniFi-Protect-1-16-0/eb8cc90d-7ecd-456f-ba98-fb6f288bc091#comment/069aae81-966a-4cbd-97ed-659a8cac6365

**Comment by AngellusMortis** [comment]
Doing more digging, I did find [.construct](https://pydantic-docs.helpmanual.io/usage/models/#creating-models-without-validation) for pydantic would allows you to make models without runtime validation.

This could be a good compromise if the API proves to be too unstable. My recommendation without knowing how often there are breaking API changes would be to let this go as is now with strict type enforcement since it is all still unused code ("beta"). See how well it plays out, i.e. when it is all done, release it as a beta version and leave it out in the wild for a while. Then if everything goes badly, I can make another PR at that point to let you choose between the type enforcement (for development) or not (for "production").

**Comment by bdraco** [comment]
Ideally we could hide the strict validation behind some type of feature flag to make it trivial to disable if we find that we need it.

**Comment by briis** [comment]
Hi both,
Overall I like what you are doing with the code, but I must also admit, that with my Python skills it is already hard for me to do additions and debug if something goes wrong with this module. I simply don't have the basic understanding that the two of you have. 
My day job is working as a Vice President for an IT Company, so whatever I do on the coding side is done out of pure interest and everything I know about Python is self taught, thus I don't have the basic structure of the language in place.

I am so happy that the two of you have chipped in, and make this module better - we could not be where we are without the effort from both of you. So with that said, I will leave the above decision to the two of you. 

The thing I can talk about is the changes I have experienced over time with the UI API. In the time I have worked with it, I have observed the following:
- The API changes often. It is almost always additions to the API like support for new devices. 
- The only real breaking change I have noticed while working with this, was when the changed the behaviour of the Motion detection. Going from always sending a motion detection to the event API, to where it is today, only send motion event when the recording mode is <> never.
- Me being located in Europe, creates some challenges, as there are new devices for Protect coming all the time, and the Early Access store here, is way behind the one in the US, meaning I cannot get my hands on the new things - and I get request all the time from users who have this new items. A Good example of that is new Smart Sensor, which has been available in the EA store in the US for month, but is still not in the EU Store. The above will not solve this issue, I just wanted to mention it, as this is a good example of a new device, that is just not plug and play to the current API and we need to structure the code so that it is easy to add, not only new devices, but also new type of devices.

So in short, this module should be able to:
- Have the ability to support new device types
- Relative easy support changes in the event detection

If this PR supports the above, then I am all for, but I cannot judge if that is the case - you two take that discussion.

One last ask - Could you either make a script that I can execute to run through all the tests, after a change, or document the steps for me - again I am not at all familiar with this. Thanks

**Comment by AngellusMortis** [comment]
So if they do not often make breaking changes it sounds like the strong type enforcement might be safe. It is just a question then if you, @briis are okay with all of the changes if you want to support the changes. Having the real classes will make it easier to build stuff on _top_ of `pyunifiprotect` so it should make the Home Assistant `unifiprotect` module easier to support. But it will make making additions to the base more complex. 

That being said, I am not going anywhere. At this point I have over $2000 of Unifi hardware so as long as I have something that is using Unifi Protect that is not Unifi Protect, even if it is not HA, I will want to help out with this project. I think at the end of the day it is your call @briis if you want to learn how to make/support the new classes I am making and you are okay with my helping out more as you learn. There are actual real classes now with inheritance and everything, so it is a lot more then just passing around some dictionaries.

As for how to run/update tests, that is pretty simple:

Make a `.env` file with your credential / login data in it:

```
UFP_USERNAME="user"
UFP_PASSWORD="password"
UFP_ADDRESS="192.168.1.1"
UFP_PORT=443
UFP_VERIFY=False
```

Then (optionally) you can generate new test data from your Protect instance. As soon as you run this command, you can trigger whatever events you want tested. I generally trigger motion for my FloodLight, camera and doorbell. Then also ring my doorbell. It is about all I can trigger in 30 seconds. Haha.

```
unifi-protect generate-sample-data
```

Finally, you run the actual tests with just

```
pytest
```

All of this is mostly optional though since everything is set up to automatically run in CI. So tests will automatically run there. I will make a separate PR now with the docs on how to generate the data and test the application.

**Comment by AngellusMortis** [comment]
PR to add more setup docs as mentioned: https://github.com/briis/pyunifiprotect/pull/47

**Comment by briis** [comment]
Thanks Christopher. Again I like the structure that is coming in to this module, and I don't mind it becoming more complex to add additions to, as long as I can ask for help and ask "dumb" questions. And as you say:

> That being said, I am not going anywhere

So I hope you (and Nick) can do the complex stuff when needed.

**Comment by AngellusMortis** [comment]
If there is a different form of communication you would like to set up more general discussion, I am open to it. I have a HA Community forum account, reddit, Discord, Signal, etc.

**Comment by briis** [comment]
As we are in different timezones I like email, so that I can respond to it when I have time. I often PM Nick on the HA Community system, so if I could get your username for that, I can use this. Mine is the same as here @briis.

Also I could add you as a co-owner to this project if you want?

**Comment by AngellusMortis** [comment]
My email is on my Github profile and HA community user name should be the same as well (@AngellusMortis).

**Comment by bdraco** [comment]
I'm also frequently on Home Assistant discord as bdraco#6479

**Comment by AngellusMortis** [comment]
I generally have Discord open like all of the time. Looks like I am still in the HA discord. I am terrible at mingling though (introvert) so I generally just lurk in Discords and do not talk to anyone. haha. `Angellus Mortis#1769`

**Comment by AngellusMortis** [comment]
Okay, I should have all of the core functionality supported for decoding/encoding for basically all of the devices I own. The `TODOs` shows areas where there are still gaps. Most of which are largely because I either cannot figure out how to make the data into something useful (`profileImg` on `cloudAccount`) or I do not have the hardware I need to do it (ptz, sensors, door locks, etc.).

**Comment by bdraco** [comment]
It looks good.  I want to do some more testing with this tomorrow and run through again after sleep.  Out of time for today as its after midnight local time.

**Review by bdraco — CHANGES_REQUESTED** [review]
```
Logger: homeassistant.setup
Source: setup.py:179
First occurred: 4:15:39 PM (1 occurrences)
Last logged: 4:15:39 PM

Setup failed for unifiprotect: Unable to import component: No module named 'pyunifiprotect.data'
```

It looks like the `data` directory needs to be added as a package in setup.cfg

**Comment by AngellusMortis** [comment]
Nice catch. It should be fixed now.

**Comment by bdraco** [comment]
Will test again tomorrow. Ran out of time for today.

**Comment by bdraco** [comment]
With the latest change, it is installing correctly now. 👍 

Unfortunately, motion sensors aren't working.

**Comment by AngellusMortis** [comment]
I was double checking over the PR and looking. I was about to say I did not actually change anything in the existing code. Haha. Just moved it around a bit.

**Comment by AngellusMortis** [comment]
Ran into some issues with updating the models from the Websocket so I did some major refactors that I figured would be best to go into this PR since it had not been merged yet. 

I split out the pylint changes to a separate PR if you would like to merge it separately. Otherwise they are also included in this PR to make CI pass. https://github.com/briis/pyunifiprotect/pull/48

**Comment by AngellusMortis** [comment]
I went ahead and merge in my completed changes for adding the new events / websocket stuff to this PR as well since I assume @bdraco has been busy with HA. You can assume this is a "new" PR except for the conversation we already had with Pydantic. I updated the title and the body of the PR accordingly.

**Comment by bdraco** [comment]
I will come back and take another look as soon as the HA release calms down

**Comment by bdraco** [comment]
I'm seeing the log full of

```
2021-10-15 08:52:05 WARNING (MainThread) [pyunifiprotect.unifi_protect_server] Unifi OS: Websocket connection not active, failing back to polling
2021-10-15 08:52:07 WARNING (MainThread) [pyunifiprotect.unifi_protect_server] Unifi OS: Websocket connection not active, failing back to polling
2021-10-15 08:52:09 WARNING (MainThread) [pyunifiprotect.unifi_protect_server] Unifi OS: Websocket connection not active, failing back to polling
2021-10-15 08:52:11 WARNING (MainThread) [pyunifiprotect.unifi_protect_server] Unifi OS: Websocket connection not active, failing back to polling
2021-10-15 08:52:13 WARNING (MainThread) [pyunifiprotect.unifi_protect_server] Unifi OS: Websocket connection not active, failing back to polling
2021-10-15 08:52:15 WARNING (MainThread) [pyunifiprotect.unifi_protect_server] Unifi OS: Websocket connection not active, failing back to polling
2021-10-15 08:52:17 WARNING (MainThread) [pyunifiprotect.unifi_protect_server] Unifi OS: Websocket connection not active, failing back to polling
2021-10-15 08:52:19 WARNING (MainThread) [pyunifiprotect.unifi_protect_server] Unifi OS: Websocket connection not active, failing back to polling
2021-10-15 08:52:21 WARNING (MainThread) [pyunifiprotect.unifi_protect_server] Unifi OS: Websocket connection not active, failing back to polling
2021-10-15 08:52:23 WARNING (MainThread) [pyunifiprotect.unifi_protect_server] Unifi OS: Websocket connection not active, failing back to polling
2021-10-15 08:52:25 WARNING (MainThread) [pyunifiprotect.unifi_protect_server] Unifi OS: Websocket connection not active, failing back to polling
2021-10-15 08:52:27 WARNING (MainThread) [pyunifiprotect.unifi_protect_server] Unifi OS: Websocket connection not active, failing back to polling
2021-10-15 08:52:29 WARNING (MainThread) [pyunifiprotect.unifi_protect_server] Unifi OS: Websocket connection not active, failing back to polling
2021-10-15 08:52:31 WARNING (MainThread) [pyunifiprotect.unifi_protect_server] Unifi OS: Websocket connection not active, failing back to polling
2021-10-15 08:52:33 WARNING (MainThread) [pyunifiprotect.unifi_protect_server] Unifi OS: Websocket connection not active, failing back to polling
2021-10-15 08:52:35 WARNING (MainThread) [pyunifiprotect.unifi_protect_server] Unifi OS: Websocket connection not active, failing back to polling
root@mauinasa:/volume1/docker/89-homeassistant# grep 'pyunifiprotect.unifi_protect_server' home-assistant.log |wc
  10487  157272 1530998
```

**Comment by bdraco** [comment]
Its happens after

```
2021-10-15 02:56:03 ERROR (MainThread) [pyunifiprotect.unifi_protect_server] Error processing websocket message
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/pyunifiprotect/unifi_protect_server.py", line 349, in _setup_websocket
    self._process_ws_message(msg)
  File "/usr/local/lib/python3.9/site-packages/pyunifiprotect/unifi_protect_server.py", line 1041, in _process_ws_message
    self._process_event_ws_message(action_json, data_json)
  File "/usr/local/lib/python3.9/site-packages/pyunifiprotect/unifi_protect_server.py", line 1110, in _process_event_ws_message
    device_id, processed_event = event_from_ws_frames(
  File "/usr/local/lib/python3.9/site-packages/pyunifiprotect/unifi_data.py", line 430, in event_from_ws_frames
    state_machine.add(event_id, data_json)
  File "/usr/local/lib/python3.9/site-packages/pyunifiprotect/unifi_data.py", line 736, in add
    self._events[event_id] = event_json
  File "/usr/local/lib/python3.9/site-packages/pyunifiprotect/data/types.py", line 20, in __setitem__
    self.popitem(False)
TypeError: dict.popitem() takes no arguments (1 given)
```

**Comment by bdraco** [comment]
I deployed the latest to all my production instances.

Will report back if anything comes up

**Comment by bdraco** [comment]
After running for a while, hit another case

```
root@alexanderalpha:/volume1/docker/alexander-homeassistant# grep 'pyunifiprotect.unifi_protect_server' home-assistant.log |wc
  40825  612261 5960034
```

```
2021-10-16 23:49:51 ERROR (MainThread) [pyunifiprotect.unifi_protect_server] Error processing websocket message
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/pyunifiprotect/unifi_protect_server.py", line 349, in _setup_websocket
    self._process_ws_message(msg)
  File "/usr/local/lib/python3.9/site-packages/pyunifiprotect/unifi_protect_server.py", line 1041, in _process_ws_message
    self._process_event_ws_message(action_json, data_json)
  File "/usr/local/lib/python3.9/site-packages/pyunifiprotect/unifi_protect_server.py", line 1110, in _process_event_ws_message
    device_id, processed_event = event_from_ws_frames(
  File "/usr/local/lib/python3.9/site-packages/pyunifiprotect/unifi_data.py", line 430, in event_from_ws_frames
    state_machine.add(event_id, data_json)
  File "/usr/local/lib/python3.9/site-packages/pyunifiprotect/unifi_data.py", line 736, in add
    self._events[event_id] = event_json
  File "/usr/local/lib/python3.9/site-packages/pyunifiprotect/data/types.py", line 21, in __setitem__
    self.popitem(False)
KeyError: 'dictionary is empty'
```

**Comment by AngellusMortis** [comment]
Updates from beta5.

* Adds 2K option `max_camera_capacity`.

(not related to the beta, just things updates from testing against real data):

* Adds `videoExported` event
* Rounds timedelta so help with weird float math causing `int` to casting to a different value then what was originally passed in.

**Comment by AngellusMortis** [comment]
Pretty sure those should be unrelated. https://github.com/briis/pyunifiprotect/pull/66 has all of the actual changes that affect the code that runs in HA and I am not seeing anything in there that would affect how many connections it makes to Protect. 

Either PR is good to merge in either order. This one is just 66 + the data models.

**Comment by bdraco** [comment]
Saw 

```
2021-10-30 14:32:11 WARNING (MainThread) [pyunifiprotect.unifi_protect_server] Unifi OS: Websocket connection not active, failing back to polling
2021-10-30 14:32:14 WARNING (MainThread) [pyunifiprotect.unifi_protect_server] Unifi OS: Websocket connection not active, failing back to polling
2021-10-30 14:32:15 WARNING (MainThread) [pyunifiprotect.unifi_protect_server] Unifi OS: Websocket connection not active, failing back to polling
2021-10-30 14:32:17 WARNING (MainThread) [pyunifiprotect.unifi_protect_server] Unifi OS: Websocket connection not active, failing back to polling
2021-10-30 14:32:19 WARNING (MainThread) [pyunifiprotect.unifi_protect_server] Unifi OS: Websocket connection not active, failing back to polling
2021-10-30 14:32:21 WARNING (MainThread) [pyunifiprotect.unifi_protect_server] Unifi OS: Websocket connection not active, failing back to polling
2021-10-30 14:32:23 WARNING (MainThread) [pyunifiprotect.unifi_protect_server] Unifi OS: Websocket connection not active, failing back to polling
2021-10-30 14:32:25 WARNING (MainThread) [pyunifiprotect.unifi_protect_server] Unifi OS: Websocket connection not active, failing back to polling
2021-10-30 14:32:27 WARNING (MainThread) [pyunifiprotect.unifi_protect_server] Unifi OS: Websocket connection not active, failing back to polling
2021-10-30 14:32:29 WARNING (MainThread) [pyunifiprotect.unifi_protect_server] Unifi OS: Websocket connection not active, failing back to polling
```

Right before I restarted.

```
2021-10-30 00:36:57 ERROR (MainThread) [pyunifiprotect.unifi_protect_server] Error processing websocket message
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/pyunifiprotect/unifi_protect_server.py", line 282, in _setup_websocket
    self._process_ws_message(msg)
  File "/usr/local/lib/python3.9/site-packages/pyunifiprotect/unifi_protect_server.py", line 949, in _process_ws_message
    self._process_event_ws_message(action_json, data_json)
  File "/usr/local/lib/python3.9/site-packages/pyunifiprotect/unifi_protect_server.py", line 1018, in _process_event_ws_message
    device_id, processed_event = event_from_ws_frames(
  File "/usr/local/lib/python3.9/site-packages/pyunifiprotect/unifi_data.py", line 466, in event_from_ws_frames
    state_machine.add(event_id, data_json)
  File "/usr/local/lib/python3.9/site-packages/pyunifiprotect/unifi_data.py", line 772, in add
    self._events[event_id] = event_json
  File "/usr/local/lib/python3.9/site-packages/pyunifiprotect/data/types.py", line 19, in __setitem__
    self.popitem(False)
KeyError: 'dictionary is empty'
```

**Comment by bdraco** [comment]
@AngellusMortis Would you mind fixing the conflicts?

I'll pull and update my instances for additional feedback after

**Comment by AngellusMortis** [comment]
You mentioned that stacktrack before. It has got to be unrelated since `FixSizeOrderedDict` was never actually changed. I can make a separate PR to try to fix that though. And yeah, updating the branch now.

**Comment by bdraco** [comment]
> You mentioned that stacktrack before. It has got to be unrelated since `FixSizeOrderedDict` was never actually changed. I can make a separate PR to try to fix that though. And yeah, updating the branch now.

Agreed

It's probably the cause of the websocket not working reports. 🤞

**Comment by AngellusMortis** [comment]
As you can see those, there are no _zero_ changes to the `UpvServer` class (https://github.com/briis/pyunifiprotect/pull/46/files#diff-deca4cd99e889cb04c9ad396127639506a3768d5ba5502975108ec493ec14266).

All of the data models stuff requires you to opt into using the new `ProtectAPIClass`. `utils.py` is only used by the new code I have been writing. I have no refactored it into any of the existing to code to keep it separate.

**Review by bdraco — APPROVED** [review]

**Comment by AngellusMortis** [comment]
> Agreed

Reproduced the issue in tests. I can definitely fix it in a separate PR. Looks like the "fixed size" part was just never working. lol
